compiler/hash-typecheck/src/storage/primitives.rs

```diff
@@ -0,0 +1,514 @@
+#[derive(Debug, Clone)]
+pub struct TrtDef {
+    pub name: Option<Identifier>,
+    pub members: Members,
+}
+
+/// A nominal definition, which for now is either a struct or an enum.
+#[derive(Debug, Clone)]
+pub enum NominalDef {
+    Struct(StructDef),
+    Enum(EnumDef),
+}
+
+/// A tuple type, containg parameters as members.
+#[derive(Debug, Clone)]
+pub struct TupleTy {
+    pub members: Params,
+}
+
+/// A function type, with a set of input parameters and a return type.
+///
+/// All the parameter types and return type must be level 0
+#[derive(Debug, Clone)]
+pub struct FnTy {
+    pub params: Params,
+    pub return_ty: TyId,
+}
+
+/// A type function type.
+///
+/// A type function is a compile-time function that works on types. Type function return types can
+/// be level 0, level 1 or level 2. It has a general set of "base" parameters and return type.
+///
+/// These are refined in the `cases` field, which provides conditional values for the return value
+/// of the function, based on what the arguments are.
+///
+/// For example, consider:
+///
+/// ```ignore
+/// Dog := <T> => struct(foo: T);
+///
+/// Dog ~= <T: Hash> => impl Hash {
+///     hash = (self) => self.foo.hash();
+/// };
+///
+/// Dog ~= <T: Hash ~ Eq> => impl FindInHashMap {
+///     ...
+/// };
+/// ```
+///
+/// Then, the definition of the type function `Dog` would look something like:
+///
+/// ```ignore
+/// TyFnTy {
+///     general_params = (T: Ty::Ty = Value::Unset),
+///     general_return_ty = Ty::Ty,
+///     cases = {
+///         (T: Ty::Ty) -> Ty::Ty => Value::NominalDef(DogStructDef),
+///         (T: Ty::Ty(HashTraitDef)) -> Ty::Ty(HashTraitDef) => Value::Merge([
+///             Value::NominalDef(DogStructDef),
+///             Value::Mod(
+///                 origin=TraitImpl(Value::Trt(HashTraitDef)),
+///                 members=..
+///             ),
+///         ]),
+///         (T: Ty::Merge([Ty::Ty(HashTraitDef), Ty::Ty(EqTraitDef)]))
+///             -> Ty::Ty(FindInHashMapTraitDef) =>
+///             => Value::Merge([
+///                 Value::NominalDef(DogStructDef),
+///                 Value::Mod(
+///                     origin=TraitImpl(Value::Trt(FindInHashMapTraitDef)),
+///                     members=..
+///                 ),
+///             ])
+///     }
+/// }
+/// ```
+///
+/// At any point, the resolved type of `Dog<T>` is the merged type of the return type of each case
+/// which matches `T`. In other words, cases are not short-circuiting; they are all evaluated and
+/// then combined.
+///
+/// The `general_return_ty` field is always a supertype of the return type of each case.
+#[derive(Debug, Clone)]
+pub struct TyFnValue {
+    /// An optional name for the type function, if it is directly assigned to a binding.
+    pub name: Option<Identifier>,
+    pub general_params: Params,
+    pub general_return_ty: TyId,
+    pub cases: Vec<TyFnCase>,
```

Comment:
This is a pretty interesting way to store what are essentially implementations of this type function